components/utils.py

```diff
@@ -1223,3 +1223,79 @@ def verify_service_http(service_name, url, predicate=None):
         return check_http_response(url, predicate)
     except (IOError, ValueError) as e:
         ctx.abort_operation('{0} error: {1}: {2}'.format(service_name, url, e))
+
+
+def validate_upgrade_directories(service_name):
+    try:
+        ctx_factory.get(service_name)
+    except IOError:
+        ctx.abort_operation('Service {0} has no properties file'.format(
+            service_name))
+
+    if os.path.exists(
+            ctx_factory.get_rollback_properties_dir(service_name)):
+        ctx.abort_operation('Rollback properties directory exists for '
+                            'service {0}'.format(service_name))
+
+    if not os.path.exists(
+            resource_factory.get_resources_dir(
+                service_name)):
+        ctx.abort_operation('Resources directory does not exist for '
+                            'service {0}'.format(service_name))
+
+    if os.path.exists(
+            resource_factory.get_rollback_resources_dir(
+                service_name)):
+        ctx.abort_operation('Rollback resources directory exists for '
+                            'service {0}'.format(service_name))
+
+
+def parse_jvm_heap_size(heap_size):
+    if heap_size.endswith('g'):
+        multiplier = 10**3
+    elif heap_size.endswith('m'):
+        multiplier = 1
+    else:
+        raise ValueError(heap_size)
+    return int(heap_size[:-1]) * multiplier
+
+
+def changed_upgrade_properties(service_name):
+    """Delta of the service's upgrade and install properties.
+
+    Look up the upgrade and install properties for the service, return a dict
+    of {property_name: (original_value, upgrade_value)}
+    """
+    install_properties = ctx_factory.get(service_name)
+    upgrade_properties = ctx.node.properties.get_all()
+    changed = {}
+    for property_name, original_value in install_properties.items():
+        changed_value = upgrade_properties.get(property_name)
+        if original_value != changed_value:
+            changed[property_name] = (original_value, changed_value)
+    return changed
+
+
+def verify_immutable_properties(service_name, properties):
```

Comment:
It now short-circuits and returns empty, if `use_existing_on_upgrade` is true. Github doesn't show it in the shortened diff here (but it does in the full diff)